<img src="https://coursesandconferences.wellcomeconnectingscience.org/wp-content/themes/wcc_courses_and_conferences/dist/assets/svg/logo.svg" width="300" height="50"> 

# Viral Genomics and Bioinformatics Asia 2024

[Wellcome Connecting Science GitHub Home Page](https://github.com/WCSCourses) <br /> 
[Wellcome Connecting Science Website](https://coursesandconferences.wellcomeconnectingscience.org/)

Hands-on training in viral genome sequence analysis and interpretation of large-scale sequencing genomics data

## Course overview

Viruses contribute to a significant public health burden globally. In Asia, viral infections such as influenza, dengue fever, hand, foot and mouth disease, chikungunya, and viral hepatitis kill hundreds of thousands of people each year.  Besides health issues, emerging viral illnesses and epidemics have a global economic impact. According to the World Bank, the SARS-CoV-2 pandemic caused a more than $10 trillion deficit in the global economy in 2020-2021 alone. This demonstrates the impact of viral illnesses not only on human and animal health but also on the social and economic growth of the world.

Technological advances in science and research are making it possible for rapid detection, surveillance and management of viral diseases and outbreaks. In recent years, next-generation sequencing (NGS) technologies have played an important role in the identification and classification of viruses, in the detection of drug-resistance mutations and the treatment and surveillance of viral diseases. Early identification of a virus and rapid characterisation of its genome will aid towards better treatment and help in controlling the disease spread. Along with the advances in sequencing technologies, it is necessary to develop data management methods, computational analysis tools and human expertise to cope with the sequence deluge. The number of researchers able to interpret NGS sequence data is currently limited.   

The Viral Genomics and Bioinformatics course will teach participants how to analyse viral genome sequences and interpret genomic data obtained by high-throughput sequencing. The use of the command line to analyse viral NGS data, quality control of sequences, reference mapping and de novo assembly of viral genomes, pathogen identification from metagenomics data, and creating phylogenetic trees will be covered in the course.

**Course Runs**      

| Course Date | Course Title | Location |Citation DOI |
|-------------|--------------|----------|-------------|
| 12 to 17 May 2024  | [Viral Genomics and Bioinformatics](https://github.com/WCSCourses/Viral_Genomics_and_Bioinfo) | OUCRU, Vietnam |[![DOI](https://zenodo.org/badge/DOI/10.5281/zenodo.14038520.svg)](https://doi.org/10.5281/zenodo.14038520) |

******

### Citing and Re-using Course Material

The course data are free to reuse and adapt with appropriate attribution. All course data in these repositories are licensed under the <a rel="license" href="https://creativecommons.org/licenses/by-nc-sa/4.0/">Attribution-NonCommercial-ShareAlike 4.0 International (CC BY-NC-SA 4.0)</a>. <a rel="license" href="http://creativecommons.org/licenses/by/4.0/"><img alt="Creative Commons Licence" style="border-width:0" src="https://i.creativecommons.org/l/by-nc-sa/4.0/88x31.png" /></a><br /> 

Each course landing page is assigned a DOI via Zenodo, providing a stable and citable reference. These DOIs can be found on the respective course landing pages and can be included in CVs or research publications, offering a professional record of the course contributions.

### Interested in attending a course?

Take a look at what courses are coming up at [Wellcome Connecting Science Courses & Conference Website](https://coursesandconferences.wellcomeconnectingscience.org/our-events/).

---
